turbopack/crates/turbopack-node/js/src/transforms/webpack-loaders.ts

```diff
@@ -159,6 +163,22 @@ type ResolveOptions = {
   importFields?: string[];
 };
 
+// Patch process.env to track which env vars are read
+const originalEnv = process.env;
+const readEnvVars = new Set<string>();
```

Comment:
- Some environment variables are used by nodejs itself (mostly these start with a `NODE_` prefix).
- Native extensions can bypass the `process.env` object, calling `getenv` directly.
- Child processes spawned by the webpack loader will receive all environment variables.

A couple options that could make this more robust:

- (Most robust) Sanitize environment variables before running node, and pass in `originalEnv` to the node process ourselves as a serialized JSON object. Native extensions and child processes won't have access to the full set of environment variables, but we'll have 100% accurate tracking. If we do this, we might need an option to let users opt-out of this behavior.

- (Less likely to break code that depends on this) Mark environment variables starting with `NODE_` as always depended on. We can make an additional "always depended on" allowlist for known loaders that use child processes or native extensions that read environment variables.

If we go with the sanitization approach, we could avoid making it a breaking change by only turning on sanitization and more granular tracking when persistent caching is enabled.

We could talk with the turborepo folks to figure out how they deal with similar problems. I think they do sanitization with an optional opt-out.

![Screenshot 2024-11-25 at 10.00.41 AM.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/HAZVitxRNnZz8QMiPn4a/53ac0d13-4c7d-4312-8ed8-0b2030b85129.png)